File: Mar 13 20. Valid Parentheses (Easy)/Solution.py
class Solution(object):
    def isValid(self, s):
        # Creat stack and opening and closing hashmap
        stack = []
        opening = set(['(', '{', '['])
        closing = {'}':'{', ']':'[', ')':'('}
        
        
        for char in s:
            # Append opening parentheses to stack
            if char in opening:
                stack.append(char)
            # Pop parentheses once find a closing parentheses
            elif char in closing:
                # If no corresponding parentheses return False
                if not stack or closing[char] != stack[-1]:
                    return False
                stack.pop()
                
        return not stack

File: Mar 13 20. Valid Parentheses (Easy)/Solution.go
func isValid(s string) bool {
    // Creat stack and opening and closing hashmap
    stack := make([]rune, 0)
    opening := map[rune]rune{
        '(' : ')',
        '[' : ']',
        '{' : '}',
    }
    closing := map[rune]rune{
        ')' : '(',
        ']' : '[',
        '}' : '{',
    }
    
    for _, char := range s{
        // Append opening parentheses to stack
        if _, found := opening[char]; found{
            stack = append(stack, char)
        // Pop parentheses once find a closing parentheses
        } else if _, found := closing[char]; found{
            // If no corresponding parentheses return False
            if len(stack) == 0 || stack[len(stack) - 1] != closing[char]{
                return false
            }
            stack = stack[:len(stack) - 1]
        }
    }
    return len(stack) == 0
}
